src/drivers/fluid_wasapi.c

```diff
@@ -0,0 +1,811 @@
+
+    dev = FLUID_NEW(fluid_wasapi_audio_driver_t);
+
+    if(dev == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: out of memory.");
+        return NULL;
+    }
+
+    FLUID_MEMSET(dev, 0, sizeof(fluid_wasapi_audio_driver_t));
+    
+    dev->func = func;
+    dev->user_pointer = data;
+    dev->buffer_duration = periods * period_size / sample_rate;
+    dev->channels_count = audio_channels * 2;
+    dev->float_samples = (wfx.Format.wFormatTag == WAVE_FORMAT_IEEE_FLOAT);
+    buffer_duration_reftime = (long long)(dev->buffer_duration * 1e7 + .5);
+    periods_reftime = (long long)(period_size / sample_rate * 1e7 +.5);
+
+    ret = CoInitializeEx(NULL, COINIT_MULTITHREADED);
+    if(FAILED(ret))
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: cannot initialize COM. %x", (unsigned)ret);
+        goto cleanup;
+    }
+
+    ret = CoCreateInstance(
+            &_CLSID_MMDeviceEnumerator, NULL,
+            CLSCTX_ALL, &_IID_IMMDeviceEnumerator,
+            (void**)&denum);
+    if(FAILED(ret))
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: cannot create device enumerator. %x", (unsigned)ret);
+        goto cleanup;
+    }
+
+    if(fluid_settings_dupstr(settings, "audio.wasapi.device", &dname) == FLUID_OK)
+    {
+        mmdev = fluid_wasapi_find_device(denum, dname);
+        if (mmdev == NULL)
+        {
+            goto cleanup;
+        }
+    }
+    else
+    {
+        ret = IMMDeviceEnumerator_GetDefaultAudioEndpoint(
+                denum,
+                eRender,
+                eMultimedia,
+                &mmdev);
+        if(FAILED(ret))
+        {
+            FLUID_LOG(FLUID_ERR, "wasapi: cannot get default audio device. %x", (unsigned)ret);
+            goto cleanup;
+        }
+    }
+
+    FLUID_FREE(dname);
+    
+    ret = IMMDevice_Activate(mmdev,
+            &_IID_IAudioClient,
+            CLSCTX_ALL, NULL,
+            (void**)&dev->aucl);
+    if(FAILED(ret))
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: cannot activate audio client. %x", (unsigned)ret);
+        goto cleanup;
+    }
+    
+    share_mode = AUDCLNT_SHAREMODE_SHARED;
+    if(fluid_settings_getint(settings, "audio.wasapi.exclusive-mode", &i) == FLUID_OK)
+    {
+        if(i)
+        {
+            share_mode = AUDCLNT_SHAREMODE_EXCLUSIVE;
+        }
+    }
+    if(share_mode == AUDCLNT_SHAREMODE_SHARED)
+    {
+        FLUID_LOG(FLUID_INFO, "wasapi: using shared mode.");
+        periods_reftime = 0;
+    }
+    else
+    {
+        FLUID_LOG(FLUID_INFO, "wasapi: using exclusive mode.");
+    }
+
+    ret = IAudioClient_IsFormatSupported(dev->aucl, share_mode, (const WAVEFORMATEX*)&wfx, (WAVEFORMATEX**)&rwfx);
+    if(FAILED(ret))
+    {
+        FLUID_LOG(FLUID_ERR, "wasapi: device doesn't support the mode we want. %x", (unsigned)ret);
+        goto cleanup;
+    }
+    else if(ret == S_FALSE)
+    {
+        FLUID_LOG(FLUID_INFO, "wasapi: requested mode cannot be fully satisfied.");
```

Comment:
This log level should be kept at `FLUID_INFO`. If it is supposed to occur very rarely, perhaps even promote it to `FLUID_WARN`.